### ISSUE TITLE
Regression in order of onsave related execution

### ISSUE THREAD

**Issue by gentlecolts** [Yellow-Dog-Man/Resonite-Issues#2421]
### Describe the bug?

One of my personal protection tools scrambles the names of proxies to prevent abuse via other users manipulating them without my consent.  This is on a toggle, and not permanent, however the new names are randomly generated.  One feature I was previously relying on is that, on save, my system would un-scramble the slot names, as those original names were necessary to detect proxies in the first place.

I don't know exactly when it started, but resonite no longer executes in the same order, and while the system will correctly un-scramble names on save, the scrambled names are what gets saved

### To Reproduce

Stated above, but I can demonstrate this ingame if needed.  The tool is in my public folder as well

### Expected behavior

Resonite should evaluate slot names/states AFTER all save events have been processed

### Resonite Version Number

Beta 2024.6.25.1149

### What Platforms does this occur on?

Windows, Linux

### Log Files

n/a

### Reporters

zangooseoo

**Comment by gentlecolts** [comment]
For clarity, I am not reporting about a bug in my system, we do not need to discuss that and I'm considering revisions to it already.  This ticket is specifically about the seeming change event execution order

**Comment by shiftyscales** [comment]
Please include a sample node graph that exhibits the issue you are highlighting, and attach it as a ResonitePackage, or link to the object / a public folder containing that object, or in a world containing that object.

Please also include supplemental material (screenshots / log files) that visualize the expected vs actual results.

**Comment by gentlecolts** [comment]
Pasting it doesn't really do anything so I'm not sure if this is right but hitting copy url on the tool in question gives me the following: `resrec:///U-zangooseoo/R-50fbfcbd-0570-450b-b102-018ff324990e`

If that does not work, my public folder is at `resrec:///U-zangooseoo/R-bd44fba0-395f-4c88-b8bf-bf3545b22690` and it can be found at the root, labeled as "Anti-Parent by zangooseoo"

intended use:
1. spawn the object
2. select object, parent it under an avatar somewhere
3. equip avatar
4. you will now have a menu labeled "Anti-Parent", one of the options is "Rename Proxies"

This all seems to work for me, as it has for years, but the problems now arise when saving the avatar:  If rename proxies is enabled, it will not do so on the saved version, any references it was keeping to the proxy slots will break, and the feature will be broken on both the newly saved version (with proxies stuck having scrambled names) and your currently equipped version (with now unscrambled names).

This was not previously an issue, and I've made no changes to the tool, though I don't know how long this behavior has been broken

**Comment by mpmxyz** [comment]
I got a bit curious about this but this project is quite complex. I.e. as an outsider it is hard to reason about what is supposed to happen because it has many independent impulse triggering nodes.
Were you relying on the order of multiple `On Loaded` firing? That order may be unspecified. Or is it?
Which chain of events happenes vs. what is supposed to happen? I know from experience that it is much easier to find an issue with good directions.

**Comment by gentlecolts** [comment]
> Were you relying on the order of multiple `On Loaded`/`On Saved` firing

No, I am relying on the state of a hierarchy, when saved, to be evaluated after any onsaved events have fired and any changes therein have taken effect.  This behavior changed at some point and may have implications beyond the scope of this one item.

The structure/design of my project itself, as stated, is not what the point of this ticket, and I would like to avoid turning this ticket into a discussion of it.  If you are curious about what I am doing, all of the relevant behavior exists under the "proxy shuffler" slot, and most of the node count there is just hacking around the current lack of arrays.

**Comment by mpmxyz** [comment]
You missed my point but I also wasn't good at getting it across:
I'm only interested in the root problem and not at the project at all. Unfortunately the project is very big and I couldn't replicate the issues using the given instructions.
There are ways to improve the likelyhood of others understanding the issue:
- Narrow down where the actual problem is! (remove unnecessary stuff, simplify code etc. while making sure it still happens)
- Make sure the instructions are precise enough that you are guaranteed to run into the problem! If you have the issue and I don't then we did something different.

Without proper ways to replicate the issue one can only guess. I did some digging:
- What I found out is that the tool only works after you spawned an avatar which has it already built-in. Just inserting doesn't work because it never fills the proxy variables.
- After that everything worked fine.
- There also isn't any code to revert the changes that have been done during `OnSaved` which means that the proxies are reverted to normal until you toggle the state back and forth.

Here is a minimalistic example I built when I read the part "Expected Behaviour":
- Spawn `resrec:///U-TheAutopilot/R-858ce6ef-e54b-4fa4-9c92-c8b292735ae7`!
- Click the button! (It should say "Nothing".)
- Save it! It should now contain "OnSaved".
- Load it! It should now contain "OnSaved" and "OnLoaded".
As you can see this did not confirm an issue.
Note: This is no proof that there are no errors but it demonstrates that if there is one it is not just a simple "Effects of 'On Saved' are ignored for saved object".

**Comment by gentlecolts** [comment]
It's been a while since I worked on this, definitely forgot some of the quirks of it, but yeah: a) it doesn't first take effect until loaded, and b) i never re-shuffle after saving, looking again those are both correct, something to fix later.

This block of code, however, should be all that's relevant to this issue:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/2545023/1c5bdb0a-a8ff-439c-953a-f0cd6fb2efe7)

Little surprised your example didn't replicate the issue I'm seeing, perhaps it's something specific to the "save avatar" button?

**Comment by mpmxyz** [comment]
resrec:///U-TheAutopilot/R-bd03c314-9e2c-4132-b882-b6b0e575bcfc is an avatar with the button. That worked.
Can I meet you within Resonite so we can boil down the cause of this behaviour? (I'll contact you.)
![grafik](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/8538547/95989fd3-169b-4385-a85d-584f1b2cc761)

**Comment by gentlecolts** [comment]
Sorry for the delay, busy weekend, and on that note I'm not really sure what my online availability is, and I'm not sure what I can contribute for reproduction/troubleshooting other than what's been posted so far, but I'm typically on in the evenings (USA eastern) most days, and my username is in the OP for this ticket

**Comment by gentlecolts** [comment]
Is more information still needed?

**Comment by mpmxyz** [comment]
If I were to guess I'd say that this issue is still too unspecific/complex to boil down if either your setup or something within Resonite is wrong.
Have you tried applying your tool to a previously unmodified avatar? Or did that also break? (It didn't for me.)

**Comment by shiftyscales** [comment]
> Can I meet you within Resonite so we can boil down the cause of this behaviour? (I'll contact you.)

Has this happened @gentlecolts @mpmxyz? From a re-read of the above comments it still didn't sound like there was a definitive / identified cause for what causes the break, as previous attempts at producing simplified replication cases didn't sound to isolate the issue?

Having exact replication steps / a simplified replication object would help a lot- or if the more complex example is needed to cause the issue- stripping it down as much as possible to its simplest form that still causes the issue.

**Comment by shiftyscales** [comment]
Oh- so that avatar exhibits the issue then, @gentlecolts? I interpreted "that worked" as "this did not have the issue".

**Comment by gentlecolts** [comment]
Oh that's a good point, I guess we're interpreting "that worked" differently, as I took that to mean it successfully reproduced the issue.  I'll try to take a closer look at it again tonight.

**Comment by gentlecolts** [comment]
So I tried to make a reproduction for this last night and I'm kind of baffled by it.  I've done multiple approaches that're basically the exact same node flow and was not able to reproduce it, yet the tool of mine which prompted this ticket still exhibits this behavior 100% of the time

**Comment by shiftyscales** [comment]
Is your existing tool a migrated object that might have old versions of various components, etc. on it @gentlecolts? Does the issue occur with that tool if you save a new copy of it to your inventory and try to reproduce the issue using that?

**Comment by gentlecolts** [comment]
> Is your existing tool a migrated object that might have old versions of various components

Sorry for the delayed response, this is in fact a migrated item, however it has been re-saved multiple times by now if that makes any difference.

**Comment by shiftyscales** [comment]
And the issue still occurs even after it has been resaved, @gentlecolts? As mentioned- if it were a migrated item, it having to update / upgrade some of the nodes / components present could possibly cause anomalous behaviour- but even then- I'd expect that the behaviour should cease after the object is re-saved.

And I'd also expect the behaviour of migrated nodes to function the same as a freshly made replication object assuming all else is equal.

It'd still be helpful to try and isolate what is different about that specific tool / why you're unable to make a reproduction item that exhibits the same behaviour.

Otherwise, if you're definitely not able to further simplify / identify why it's behaving the way it is- I'd say just include another link to it / upload a ResonitePackage for it and the issue could be looked into further when our coders have time for it- but having a definitive / highly likely identified cause for how to reproduce it would make it easier to resolve.

**Comment by gentlecolts** [comment]
> And the issue still occurs even after it has been resaved, @gentlecolts?

yep

> And I'd also expect the behaviour of migrated nodes to function the same as a freshly made replication object assuming all else is equal.

Same, and I can't even seem to reproduce this behavior with fresh versions of what seems like the exact same node flow.  I'll pretty openly admit that the tool itself is years old and not the best-designed anyways, and my plans for a v2 would remove this issue entirely.  The fact remains though that I can reliably show this misbehavior, even if I can't replicate it, and the fact that this issue has arisen at all is pretty bizarre.  I'm kind of hoping this is just some really dumb oversight on my part but it's definitely a bother that there'd be this kind of issue at all.

I'll try to get a resonitepackage version on this thread soon

**Comment by gentlecolts** [comment]
ugg I haven't had any time to think about this and i kind of gave up on it :/
Last i looked i could not repro with any newly made objects, but could reliably do so with the migrated one in my inventory as described above.  might be a migration bug or something, who knows, i figured if i cared enough i'd remake the system anyways
